scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala

```diff
@@ -0,0 +1,140 @@
+package com.twitter.scalding.spark_backend
+
+import com.twitter.scalding.{ExecutionCounters, StatKey}
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.util.AccumulatorV2
+
+import scala.collection.mutable.{Map => MutableMap}
+import scala.collection.immutable.{Map => ImmutableMap}
+
+object SparkCounters {
+  def withRandomContextPrefix(sparkSession: SparkSession): SparkCounters =
+    new SparkCounters(sparkSession, java.util.UUID.randomUUID.toString)
+}
+
+/**
+ * We use a new SparkCounters instance for each SparkWriter to keep track of scalding counters
+ */
+class SparkCounters(sparkSession: SparkSession, counterPrefix: String) {
+
+  /**
+   * This counter is specific to the spark driver. Executors will serialize SparkCountersInternal but it will
+   * be a stateless instance obtained through the copyAndReset method. You can read the internal
+   * implementation here:
+   * https://github.com/apache/spark/blob/87c744b60507f82e1722f1488f1741cb2bb8e8e5/core/src/main/scala/org/apache/spark/util/AccumulatorV2.scala#L165
+   */
+  @transient private val accumulator: SparkCountersInternal = new SparkCountersInternal
+
+  /**
+   * This should only be called inside of SparkWriter execution.
+   */
+  def getAccumulator: SparkCountersInternal = accumulator
+
+  /**
+   * Register for spark to be aware of accumulator. Note that accumulators are now automatically garbage
+   * collected: https://github.com/apache/spark/pull/4021 so we don't need to explicitly release them.
+   */
+  sparkSession.sparkContext.register(accumulator, counterPrefix)
+
+  /**
+   * We return an ExecutionCounters interface. The results will not be stable unless spark execution has
+   * finished.
+   */
+  def asExecutionCounters(): ExecutionCounters = {
+    val copyState = MutableMap[StatKey, Long]()
+    for (kvPair <- accumulator.value) {
+      val keyEntry = StatKey(kvPair._1._2, kvPair._1._1)
+      copyState(keyEntry) = kvPair._2
+    }
+    val immutableCopyState = copyState.toMap
+    new ExecutionCounters {
+      def keys = immutableCopyState.keySet
+      def get(key: StatKey) = immutableCopyState.get(key)
+    }
+  }
```

Comment:
I believe I located the problem with the failing test. We need to reset the counter state before we return from here. I think what's going on is that flapMapping the execution re-uses the same writer which duplicates the counter values.

I confirmed that the actual counter incrementation occurs only once by stepping through breakpoints and checking on the counter state.